flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonRowDataDeserializationSchema.java

```diff
@@ -0,0 +1,492 @@
+import static org.apache.flink.formats.json.TimeFormats.RFC3339_TIMESTAMP_FORMAT;
+import static org.apache.flink.formats.json.TimeFormats.RFC3339_TIME_FORMAT;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Deserialization schema from JSON to Flink Table/SQL internal data structure {@link RowData}.
+ *
+ * <p>Deserializes a <code>byte[]</code> message as a JSON object and reads
+ * the specified fields.
+ *
+ * <p>Failures during deserialization are forwarded as wrapped IOExceptions.
+ */
+@Internal
+public class JsonRowDataDeserializationSchema implements DeserializationSchema<RowData> {
+	private static final long serialVersionUID = 8576854315236033439L;
+
+	/** Flag indicating whether to fail if a field is missing. */
+	private final boolean failOnMissingField;
+
+	/** Flag indicating whether to ignore invalid fields/rows (default: throw an exception). */
+	private final boolean ignoreParseErrors;
+
+	/** TypeInformation of the produced {@link RowData}. **/
+	private final TypeInformation<RowData> resultTypeInfo;
+
+	/**
+	 * Runtime converter that converts {@link JsonNode}s into
+	 * objects of Flink SQL internal data structures. **/
+	private final DeserializationRuntimeConverter runtimeConverter;
+
+	/** Object mapper for parsing the JSON. */
+	private final ObjectMapper objectMapper = new ObjectMapper();
+
+	/**
+	 * Creates a builder for {@link JsonRowDataDeserializationSchema}.
+	 */
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	private JsonRowDataDeserializationSchema(
+			RowType rowType,
+			TypeInformation<RowData> resultTypeInfo,
+			boolean failOnMissingField,
+			boolean ignoreParseErrors) {
+		if (ignoreParseErrors && failOnMissingField) {
+			throw new IllegalArgumentException(
+				"JSON format doesn't support failOnMissingField and ignoreParseErrors are both enabled.");
+		}
+		this.resultTypeInfo = checkNotNull(resultTypeInfo);
+		this.failOnMissingField = failOnMissingField;
+		this.ignoreParseErrors = ignoreParseErrors;
+		this.runtimeConverter = createRowConverter(checkNotNull(rowType));
+	}
+
+	@Override
+	public RowData deserialize(byte[] message) throws IOException {
+		try {
+			final JsonNode root = objectMapper.readTree(message);
+			return (RowData) runtimeConverter.convert(root);
+		} catch (Throwable t) {
+			if (ignoreParseErrors) {
+				return null;
+			}
+			throw new IOException(format("Failed to deserialize JSON '%s'.", new String(message)), t);
+		}
+	}
+
+	@Override
+	public boolean isEndOfStream(RowData nextElement) {
+		return false;
+	}
+
+	@Override
+	public TypeInformation<RowData> getProducedType() {
+		return resultTypeInfo;
+	}
+
+	// -------------------------------------------------------------------------------------
+	// Builder
+	// -------------------------------------------------------------------------------------
+
+	/**
+	 * Builder for {@link JsonRowDataDeserializationSchema}.
+	 */
+	public static class Builder {
```

Comment:
I think we will support more and more options for JSON in the future. A builder is more friendly to evolve without breaking constructors. But I'm also fine to remove builders because this is an `@Internal` class.